File: C/binary_search.c
#include <stdint.h>
#include <stdio.h>

int n = 100;

struct {
  int key;
  int data;
} table[100];

void ini_table() {
  for (int i = 0; i < n; i++) {
    table[i].key = i;
    table[i].data = i * 123;
  }
}

int binarysearch(int key) {
  int low, middle, high;
  low = 0;
  high = n - 1;

  while (low <= high) {
    middle = (low + high) / 2;
    if (key == table[middle].key) {
      return table[middle].data;
    } else if (key < table[middle].key) {
      high = middle - 1;
    } else {
      low = middle - 1;
    }
    printf("low:%d mid:%d, high:%d\n", low, middle, high);
  }
  return -1;
}

int main(void) {
  ini_table();
  int result = binarysearch(88);
  printf("%d", result);
}

File: Zig/binary_search.zig
const std = @import("std");
const print = std.debug.print;

const record = struct {
    key: i32,
    data: i32,
};

const table_len = 100;
var table: [table_len]record = undefined;

pub fn init_table() void {
    for (table[0..], 0..) |*entry, idx| {
        const key: i32 = @intCast(idx);
        entry.* = .{ .key = key, .data = key * 123 };
    }
}

pub fn binary_search(key: i32) i32 {
    var low: usize = 0;
    var high: usize = table_len - 1;

    while (low <= high) {
        const mid = (high + low) / 2;
        const mid_key = table[mid].key;

        if (key == mid_key) {
            return table[mid].data;
        } else if (key < mid_key) {
            if (mid == 0) break;
            high = mid - 1;
        } else {
            low = mid + 1;
        }
    }

    return -1;
}

pub fn main() !void {
    init_table();
    const result = binary_search(88);
    print("result:{any}\n", .{result});
}
